### ISSUE TITLE
Increment responds with "item not stored"

### ISSUE THREAD

**Issue by velimir** [memcached/memcached#957]
**Describe the bug**
*TLDR*: If running multiple requests concurrently  to increment a specific key one of the requests would return "Item not stored" response.

From what I can see it's totally possible to get `NOT_STORED` response from the increment operation according to the [code](https://github.com/memcached/memcached/blob/597645db6a2b138710f01ffe5e92e453117b987a/proto_bin.c#L333-L334). 
If I'm reading `complete_incr_bin ` correctly this could happen when one of the requests is waiting for the lock in [store_item](https://github.com/memcached/memcached/blob/597645db6a2b138710f01ffe5e92e453117b987a/proto_bin.c#L329) and the other acquires one and store the item successfully.
Once the lock is released the request that was locked attempts to store an item with `NREAD_ADD` command, which [does not store the item](https://github.com/memcached/memcached/blob/597645db6a2b138710f01ffe5e92e453117b987a/memcached.c#L1597-L1598) as it already exists.

I can't see this to be documented anywhere. Is this expected behaviour?

**WireShark dump**

```
No.     Time           Source                Src Port Destination           Dst Port Protocol Length Info
   3283 6.502757       127.0.0.1             50332    127.0.0.1             11211    MEMCACHE 156    Increment Request

Memcache Protocol, Increment Request
    Magic: Request (128)
    Opcode: Increment (5)
    Key Length: 56
    Extras length: 20
    Data type: Raw bytes (0)
    Reserved: 0
    [Value length: 0]
    Total body length: 76
    Opaque: 0
    CAS: 0
    Extras
        Amount to add: 1
        Initial value: 1
        Expiration: 600
    Key: mskd_slk_sl::89GqayaR9R2Mr6MFDSXxg9b1jQEf3RRGpoOpk3vY-Do

No.     Time           Source                Src Port Destination           Dst Port Protocol Length Info
   3284 6.502775       127.0.0.1             50333    127.0.0.1             11211    MEMCACHE 156    Increment Request

Memcache Protocol, Increment Request
    Magic: Request (128)
    Opcode: Increment (5)
    Key Length: 56
    Extras length: 20
    Data type: Raw bytes (0)
    Reserved: 0
    [Value length: 0]
    Total body length: 76
    Opaque: 0
    CAS: 0
    Extras
        Amount to add: 1
        Initial value: 1
        Expiration: 600
    Key: mskd_slk_sl::89GqayaR9R2Mr6MFDSXxg9b1jQEf3RRGpoOpk3vY-Do

No.     Time           Source                Src Port Destination           Dst Port Protocol Length Info
   3296 6.504377       127.0.0.1             11211    127.0.0.1             50333    MEMCACHE 88     Increment Response

Memcache Protocol, Increment Response
    Magic: Response (129)
    Opcode: Increment (5)
    Key Length: 0
    Extras length: 0
    Data type: Raw bytes (0)
    Status: No error (0)
    [Value length: 8]
    Total body length: 8
    Opaque: 0
    CAS: 353
    Response: 1

No.     Time           Source                Src Port Destination           Dst Port Protocol Length Info
   3297 6.504384       127.0.0.1             11211    127.0.0.1             50332    MEMCACHE 91     Increment Response (Item not stored)

Memcache Protocol, Increment Response
    Magic: Response (129)
    Opcode: Increment (5)
    Key Length: 0
    Extras length: 0
    Data type: Raw bytes (0)
    Status: Item not stored (5)
        [Expert Info (Note/Response): Increment: Item not stored]
            [Increment: Item not stored]
            [Severity level: Note]
            [Group: Response]
    [Value length: 11]
    Total body length: 11
    Opaque: 0
    CAS: 0
    Value: Not stored.

```

**To Reproduce**
Run multiple increment requests (binary proto) in parallel.

**System Information**
 - OS/Distro: [Debian based bitnami image](https://hub.docker.com/layers/bitnami/memcached/1.6.6-debian-10-r92/images/sha256-46d93d547637d3c1cd61a1419c07af04829add7ad24c202ea97520571a0836bb?context=explore)
 - Version of OS/distro: debian 10 (buster)
 - Version of memcached: 1.6.6
 - Hardware detail: image running in a minikube on a local machine (macos M1)

**Detail**
#### `stats`
```
STAT pid 1
STAT uptime 1110
STAT time 1669887504
STAT version 1.6.6
STAT libevent 2.1.8-stable
STAT pointer_size 64
STAT rusage_user 2.289417
STAT rusage_system 0.269273
STAT max_connections 1024
STAT curr_connections 152
STAT total_connections 842
STAT rejected_connections 0
STAT connection_structures 512
STAT response_obj_bytes 8176
STAT response_obj_total 7
STAT response_obj_free 6
STAT response_obj_oom 0
STAT read_buf_bytes 65536
STAT read_buf_bytes_free 49152
STAT read_buf_oom 0
STAT reserved_fds 20
STAT cmd_get 0
STAT cmd_set 4
STAT cmd_flush 0
STAT cmd_touch 0
STAT cmd_meta 0
STAT get_hits 0
STAT get_misses 0
STAT get_expired 0
STAT get_flushed 0
STAT delete_misses 0
STAT delete_hits 0
STAT incr_misses 0
STAT incr_hits 0
STAT decr_misses 0
STAT decr_hits 0
STAT cas_misses 0
STAT cas_hits 0
STAT cas_badval 0
STAT touch_hits 0
STAT touch_misses 0
STAT auth_cmds 0
STAT auth_errors 0
STAT bytes_read 13522
STAT bytes_written 12355
STAT limit_maxbytes 67108864
STAT accepting_conns 1
STAT listen_disabled_num 0
STAT time_in_listen_disabled_us 0
STAT threads 4
STAT conn_yields 0
STAT hash_power_level 16
STAT hash_bytes 524288
STAT hash_is_expanding 0
STAT slab_reassign_rescues 0
STAT slab_reassign_chunk_rescues 0
STAT slab_reassign_evictions_nomem 0
STAT slab_reassign_inline_reclaim 0
STAT slab_reassign_busy_items 0
STAT slab_reassign_busy_deletes 0
STAT slab_reassign_running 0
STAT slabs_moved 0
STAT lru_crawler_running 0
STAT lru_crawler_starts 1530
STAT lru_maintainer_juggles 2168
STAT malloc_fails 0
STAT log_worker_dropped 0
STAT log_worker_written 0
STAT log_watcher_skipped 0
STAT log_watcher_sent 0
STAT bytes 1251
STAT curr_items 5
STAT total_items 5
STAT slab_global_page_pool 0
STAT expired_unfetched 0
STAT evicted_unfetched 0
STAT evicted_active 0
STAT evictions 0
STAT reclaimed 0
STAT crawler_reclaimed 0
STAT crawler_items_checked 20
STAT lrutail_reflocked 0
STAT moves_to_cold 5
STAT moves_to_warm 0
STAT moves_within_lru 0
STAT direct_reclaims 0
STAT lru_bumps_dropped 0
```

#### `stats settings`
```
STAT maxbytes 67108864
STAT maxconns 1024
STAT tcpport 11211
STAT udpport 0
STAT inter NULL
STAT verbosity 1
STAT oldest 0
STAT evictions on
STAT domain_socket NULL
STAT umask 700
STAT growth_factor 1.25
STAT chunk_size 48
STAT num_threads 4
STAT num_threads_per_udp 4
STAT stat_key_prefix :
STAT detail_enabled no
STAT reqs_per_event 20
STAT cas_enabled yes
STAT tcp_backlog 1024
STAT binding_protocol auto-negotiate
STAT auth_enabled_sasl no
STAT auth_enabled_ascii no
STAT item_size_max 1048576
STAT maxconns_fast yes
STAT hashpower_init 0
STAT slab_reassign yes
STAT slab_automove 1
STAT slab_automove_ratio 0.80
STAT slab_automove_window 30
STAT slab_chunk_max 524288
STAT lru_crawler yes
STAT lru_crawler_sleep 100
STAT lru_crawler_tocrawl 0
STAT tail_repair_time 0
STAT flush_enabled yes
STAT dump_enabled yes
STAT hash_algorithm murmur3
STAT lru_maintainer_thread yes
STAT lru_segmented yes
STAT hot_lru_pct 20
STAT warm_lru_pct 40
STAT hot_max_factor 0.20
STAT warm_max_factor 2.00
STAT temp_lru no
STAT temporary_ttl 61
STAT idle_timeout 0
STAT watcher_logbuf_size 262144
STAT worker_logbuf_size 65536
STAT resp_obj_mem_limit 0
STAT read_buf_mem_limit 0
STAT track_sizes no
STAT inline_ascii_response no
STAT ext_item_size 512
STAT ext_item_age 4294967295
STAT ext_low_ttl 0
STAT ext_recache_rate 2000
STAT ext_wbuf_size 4194304
STAT ext_compact_under 0
STAT ext_drop_under 0
STAT ext_max_frag 0.80
STAT slab_automove_freeratio 0.010
STAT ext_drop_unread no
```

**Comment by dormando** [comment]
Hey,

I did look into this briefly a few days ago but forgot to respond, sorry. It looks like this is possibly specifically if the key doesn't already exist. So if you're shipping binary protocol incr's to the same existing key it should always work, but if the key does not exist another initial request happening at the same time could fail.

I'm not sure this was known/expected behavior, but it's been a really long time and I'm unsure anymore.

In the newer meta protocol there is a similar feature (incr, seed value on miss) and I checked there and it does hold the locks properly while seeding a missed value.

I'll need to come back to this to decide on if it's possible to fix or if we should continue to gently push people away from binary protocol. Meta doesn't have much client support yet so it's possible we'll fix it.

Thanks for the report.

**Comment by xr** [comment]
Thanks @velimir for sharing, we have the similar issues on our side, @dormando any plans to fix the race condition issues in the binary protocol incr or you suggest to move towards to seek some clients that support meta protocol? Also any suggested workaround for now for this? Thank you!

**Comment by dormando** [comment]
@xr the meta protocol fixes a lot of stuff and only has four commands.. so if at all possible that's better.

I'm pretty overloaded and modifying the binary protocol code isn't always easy. I'm not sure how I would be able to prioritize that fix right now. I'll try to take a more serious look this weekend.

**Comment by xr** [comment]
@dormando thank you!

**Comment by dormando** [comment]
sorry, didn't have time last weekend. hopefully this weekend.

**Comment by xr** [comment]
Hello @dormando any chance to have a look? Thanks!

**Comment by dormando** [comment]
I haven't yet, sorry :(
